File: leetcode/cpp/PassedProblems/206-reverse-linked-list.cpp
/*
 * 206-reverse-linked-list.cpp
 *
 *  Created on: Sep 14, 2015
 *      Author: hanzhe
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <iostream>
struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
  };
using namespace std;
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
    	if(head == NULL || head->next == NULL)
    		return head;
    	ListNode * tPre = head, *tH = head->next;
    	tPre->next = NULL;
    	while(tH != NULL){
    		ListNode * tn = tH->next;
    		tH->next = tPre;
    		tPre = tH;
    		tH = tn;
    	}
    	return tPre;
    }
};

File: leetcode/python/206-reverse-linked-list.py
#!/usr/bin/env python
# encoding: utf-8


"""
Reverse a singly linked list.
"""


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None


# 4次过，不难其实
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        t = head
        re = []
        while t != None:
            re.append(t.val)
            t = t.next
        t = head; i = 0
        re.reverse()
        while t != None:
            t.val = re[i]
            t = t.next
            i += 1
        return head
